keps/sig-api-machinery/20190726-component-config-fields-should-be-optional.md

```diff
@@ -0,0 +1,197 @@
+these nilable.  kubelet uses a ComponentConfig type that is v1beta1, I propose
+not changing those at the current time, but rather proving out the approach with
+the v1alpha1 types.
+
+## Motivation
+
+I initially raised this in a [mailing list
+thread](https://groups.google.com/d/topic/kubernetes-wg-component-standard/hO7Lmi0cwQU/discussion)
+with wg-component-standard.  When trying to create componentconfig from code,
+all fields were populated, which did not really reflect the intent.  For minimal
+output, either simply for hygene or to support kustomize, we would have needed
+another type.  We would then have had to keep that type in sync with changes to
+componentconfig.  It seemed that componentconfig could better serve as that
+type.
+
+Supporting merging of configurations that better reflects intent has come up a
+few times: there have been discussions of merging fields with flags and the
+precedence when doing so, merging fields set at different levels of a hierarchy
+(e.g. cluster-level vs node-level kube-proxy flags).  Recording the distinction
+between the zero-value and not-set allows for these operations to better reflect
+intent.
+
+### Goals
+
+- Allow generation of compact componentconfig using the same types
+- Start to replace some of the "magic sentinel" values that are used to mean "not set"
+
+### Non-Goals
+
+- Deeper refactoring of component config types into more structured objects
+  (this is valuable, merely not under the umbrella of this KEP)
+
+## Proposal
+
+In v1.16 we will change the types of the fields in the kube-proxy,
+kube-scheduler, kube-controller-manager, cloud-controller-manager
+ComponentConfig so that they are all nilable.  We believe we can keep the
+version as v1alpha1, because the yaml and json form has not changed (though the
+proto would, but componentconfig is not specified using proto).  We believe we
+can keep the internal versions with non-nilable fields, and perform the
+conversion.
+
+After demonstrating this in v1.16 (ideally in v1.17) we will change the types of
+the fields in kubelet so that they are nilable.  This may require introduction
+of a v1beta2 (TBD).
+
+### User Stories
+
+#### Can marshal to yaml
+
+When I create an object of type KubeSchedulerConfiguration and marshal it to
+yaml/json, only the fields I have set appear in the output (other than "header
+fields" like Kind & APIVersion etc).
+
+#### Can set fields to zero values
+
+When I set fields to zero-values in a KubeSchedulerConfiguration (0 for
+integers, "" for strings etc), and marshal to yaml/json, those fields do appear
+in the output, reflecting my intent.  This enables other tooling such as
+kustomize to know that I set these fields.
+
+### Implementation Details/Notes/Constraints
+
+We will begin with kube-scheduler and kube-proxy, as those are the smallest;
+there is also more desire to promote those to beta.
+
+Implementation details TBD!
```

Comment:
Off the top of my head:

1. Make everything a pointer in the external type(s).
1. Update defaulting code to check nil everywhere instead of zero values (yay for consistency!).
1. Backfill zero-value defaults for previously-non-pointer fields.
1. Run codegen.
1. Run tests.

Hopefully most logic is based off the internal type and changes are minimal. Some tests that round-trip may need updates.

Might get interesting with deeply-nested structures. Best way to find out is to try it.

Reply:
Making everything a pointer is a nice thing from consistency standpoint, however it also looks annoying in Go code. It won't have much impact on users, that treat component configs as YAML templates only, but Go code vendoring users are going to be in for a hard time (or at least, if I understand correctly, for a huge use of `k8s.io/utils/pointer`).

The real problem here is tied to `omitempty` and the JSON library, which is very cautious in respecting it. In fact, in non-trivial cases, it won't respect it at all. The Go-YAML library has a `IsZeroer` interface, but that is not supported in the builtin Go JSON library.